### PR TITLE
JCLOUDS-852:Added VMImageOpertaions

### PR DESCRIPTION
**Pull request by hsbhathiya**
Virtual Machine Image Operations. 
https://msdn.microsoft.com/en-us/library/azure/dn499771.aspx

Rebased
https://github.com/jclouds/jclouds-labs/pull/144 on master

### CODE REVIEW

**Comment by ilgrosso** [comment]
```
Tests run: 98, Failures: 0, Errors: 0, Skipped: 2
[...]
[INFO] Total time: 01:24 h
```

Looks good! @nacx @andreaturli could you please review this PR?

File: azurecompute/src/main/java/org/jclouds/azurecompute/domain/VMImage.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+      public enum Caching {
+         READ_ONLY,
+         READ_WRITE,
+         NONE
+      }
+
+      public enum OSState {
+         GENERALIZED,
+         SPECIALIZED
+
```
Comment (andreaturli) [inline azurecompute/src/main/java/org/jclouds/azurecompute/domain/VMImage.java]: please remove trailing lines

File: azurecompute/src/main/java/org/jclouds/azurecompute/domain/VMImage.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+       */
+      @Nullable public abstract Caching hostCaching();
+
+      /**
+       * Specifies the state of the operating system in the image.
+       * A Virtual Machine that is fully configured and running contains a Specialized operating system.
+       * A Virtual Machine on which the Sysprep command has been run with the generalize option contains
+       * a Generalized operating system.
+       */
+
```
Comment (andreaturli) [inline azurecompute/src/main/java/org/jclouds/azurecompute/domain/VMImage.java]: remove this line

File: azurecompute/src/main/java/org/jclouds/azurecompute/domain/VMImageParams.java [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+       */
+      @Nullable public abstract Caching hostCaching();
+
+      /**
+       * Specifies the state of the operating system in the image.
+       * A Virtual Machine that is fully configured and running contains a Specialized operating system.
+       * A Virtual Machine on which the Sysprep command has been run with the generalize option contains
+       * a Generalized operating system.
+       */
+
```
Comment (andreaturli) [inline azurecompute/src/main/java/org/jclouds/azurecompute/domain/VMImageParams.java]: remove line

File: azurecompute/src/main/java/org/jclouds/azurecompute/domain/VMImageParams.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+               .eula(in.eula()).iconUri(in.iconUri()).smallIconUri(in.smallIconUri()).privacyUri(in.privacyUri());
+      }
+
+   }
+
+   public static VMImageParams create(String name, String label, String description,
+         OSDiskConfigurationParams osDiskConfiguration, List<DataVirtualHardDisk> dataDiskConfiguration,
+         String language, String imageFamily, RoleSize.Type recommendedVMSize, String eula, URI iconUri,
+         URI smallIconUri, URI privacyUri, Boolean showGui) {
+      return new AutoValue_VMImageParams(name, label, description, osDiskConfiguration, dataDiskConfiguration,
```
Comment (andreaturli) [inline azurecompute/src/main/java/org/jclouds/azurecompute/domain/VMImageParams.java]: for lists (or any other collections) in AutoValue classe, please use ImmutableList.copyOf(dataDiskConfiguration) as recommended by auto/value doc

File: azurecompute/src/main/java/org/jclouds/azurecompute/domain/VMImage.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+   @Nullable public abstract Date publishedDate();
+
+   public static VMImage create(String name, String label, String category, String description,
+         OSDiskConfiguration osDiskConfiguration, List<DataVirtualHardDisk> dataDiskConfiguration, String serviceName,
+         String deploymentName, String roleName, String location, String affinityGroup, Date createdTime,
+         Date modifiedTime, String language, String imageFamily, RoleSize.Type recommendedVMSize, Boolean isPremium,
+         String eula, URI iconUri, URI smallIconUri,
+         URI privacyUri, Date publishedDate) {
+      return new AutoValue_VMImage(name, label, category, description, osDiskConfiguration, dataDiskConfiguration,
```
Comment (andreaturli) [inline azurecompute/src/main/java/org/jclouds/azurecompute/domain/VMImage.java]: use `ImmutableList.copyOf(dataDiskConfiguration)

File: azurecompute/src/main/java/org/jclouds/azurecompute/xml/OSConfigHandler.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+   private URI mediaLink;
+   private String ioType;
+
+   private final StringBuilder currentText = new StringBuilder();
+
+   @Override
+   public VMImage.OSDiskConfiguration getResult() {
+      VMImage.OSDiskConfiguration result = VMImage.OSDiskConfiguration
+            .create(name, hostCaching, osState, os, mediaLink, logicalDiskSizeInGB, ioType);
+      return result;
```
Comment (andreaturli) [inline azurecompute/src/main/java/org/jclouds/azurecompute/xml/OSConfigHandler.java]: can we just return it directly?

**Comment by andreaturli** [comment]
@hsbhathiya I think it looks good, only minor comments from my side.

File: azurecompute/src/main/java/org/jclouds/azurecompute/features/VMImageApi.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+
+   /**
+    * The List VM Images operation retrieves a list of the VM Images from the image repository that is associated with
+    * the specified subscription.
+    */
+   @Named("ListVMImages")
+   @GET
+   @XMLResponseParser(ListVMImagesHandler.class)
+   @Produces(APPLICATION_XML)
```
Comment (nacx) [inline azurecompute/src/main/java/org/jclouds/azurecompute/features/VMImageApi.java]: If this method does not produce a body, remove this annotation.

File: azurecompute/src/main/java/org/jclouds/azurecompute/features/VMImageApi.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+         @BinderParam(VMImageParamsToXML.class) VMImageParams params);
+
+   /**
+    * The Delete VM Image operation deletes the specified VM Image from the image repository that is associated with
+    * the specified subscription.
+    */
+   @Named("DeleteImage")
+   @DELETE
+   @Path("/{imageName}")
+   @Produces(APPLICATION_XML)
```
Comment (nacx) [inline azurecompute/src/main/java/org/jclouds/azurecompute/features/VMImageApi.java]: Same comment about the produced body and the annotation.

File: azurecompute/src/test/java/org/jclouds/azurecompute/features/VMImageApiLiveTest.java [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+                        30,
+                        "Standard");
+        VMImageParams params = VMImageParams.builder().name(CREATE_IMAGE_NAME).label(CREATE_IMAGE_NAME)
+                .description(image.description()).recommendedVMSize(image.recommendedVMSize())
+                .osDiskConfiguration(osParams).imageFamily(image.imageFamily())
+                .build();
+
+        String requestId = api().create(params);
+        assertNotNull(requestId);
+        operationSucceeded.apply(requestId);
```
Comment (nacx) [inline azurecompute/src/test/java/org/jclouds/azurecompute/features/VMImageApiLiveTest.java]: Is this something that needs to be deleted at some point when the test finishes?

File: azurecompute/src/test/java/org/jclouds/azurecompute/features/VMImageApiMockTest.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+                    .build();
+
+            assertEquals(api.create(params), "request-1");
+
+            assertSent(server, "POST", "/services/vmimages", "/vmimageparams.xml");
+        } finally {
+            server.shutdown();
+        }
+    }
+}
```
Comment (nacx) [inline azurecompute/src/test/java/org/jclouds/azurecompute/features/VMImageApiMockTest.java]: Add the missing mock tests: update, delete, and delete with fallback.

**Comment by nacx** [comment]
Also add the mock and live tests for the method you added to the `VirtualMachineApi`.

File: azurecompute/src/main/java/org/jclouds/azurecompute/features/VMImageApi.java (hunk 7 above)
Comment (nacx) [inline azurecompute/src/main/java/org/jclouds/azurecompute/features/VMImageApi.java]: Does this method generate a body?

**Comment by nacx** [comment]
>Also add the mock and live tests for the method you added to the VirtualMachineApi.

This still has to be done.

Please, when addressing review commits add the changes in separate commits. Don't squash them unless requested. That makes the review much more easy!

**Comment by hsbhathiya** [comment]
@nacx
CaptureVMImage operation is added to the VirtualMachineApi becasuse of its Request URI. All the other VMImage Operations request URIs start with https://management.core.windows.net/<subscription-id>/services/vmimages. Tests for that method is in VMImgeLiveTests.

I will add the CaptureVMImage Mock test as well.

**Comment by hsbhathiya** [comment]
Ok I won't squash by now.

**Comment by hsbhathiya** [comment]
@nacx I have addressed issues.

**Comment by nacx** [comment]
Thanks @hsbhathiya! lgtm. Mind squashing the commits?
Also, while squashing, mind adding a comment in the VirtualMachineApi mock and live test classes indicating that the capture tests are in the VMImage test classes? One would expect those to be in the VirtualMachineApi test classes, so let's make sure people will find those tests and we know they're not missing :)

**Comment by hsbhathiya** [comment]
@nacx: Commented and squashed

**Comment by nacx** [comment]
Amended de commit message and pushed to [master](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/882be03e) and [1.9.x](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/6b4de9e7). Thanks @hsbhathiya!

**Pull request closed without merging**